src/internal/storage/server.go

```diff
@@ -135,6 +143,35 @@ func (s *Server) CreateFileset(server storage.Fileset_CreateFilesetServer) error
 	})
 }
 
+func (s *Server) copyFile(ctx context.Context, uw *fileset.UnorderedWriter, msg *storage.CopyFile) error {
+	id, err := fileset.ParseID(msg.FilesetId)
+	if err != nil {
+		return err
+	}
+	fs, err := s.Filesets.Open(ctx, []fileset.ID{*id})
+	if err != nil {
+		return err
+	}
+	srcPath := pfsfile.CleanPath(msg.Src)
+	dstPath := srcPath
+	if msg.Dst != "" {
+		dstPath = pfsfile.CleanPath(msg.Dst)
+	}
+	fs = fileset.NewIndexFilter(fs, func(idx *index.Index) bool {
+		return idx.Path == srcPath || strings.HasPrefix(idx.Path, fileset.Clean(srcPath, true))
+	})
+	fs = fileset.NewIndexMapper(fs, func(idx *index.Index) *index.Index {
+		idx = protoutil.Clone(idx)
+		relPath, err := filepath.Rel(srcPath, idx.Path)
+		if err != nil {
+			panic(err)
```

Comment:
I think this is fine.  It's always a bug on our end, and I don't think you can craft an input that panics here.